contrib/slurm/slurmexecutor.py

```diff
@@ -251,57 +283,121 @@ def run_slurm(benchmark, args, log_file):
+                    first_lines,
+                )
+                jobid_match = jobid_pattern.search(str(first_lines))
+                if jobid_match:
+                    jobid = int(jobid_match.group(1))
 
         seff_command = ["seff", str(jobid)]
         logging.debug(
             "Command to run: %s", " ".join(map(util.escape_string_shell, seff_command))
         )
-        result = subprocess.run(
-            seff_command,
-            stdout=subprocess.PIPE,
-            stderr=subprocess.STDOUT,
-        )
 
-    # Runexec would populate the first 6 lines with metadata
-    with open(log_file, "r+") as file:
-        content = file.read()
-        file.seek(0, 0)
-        empty_lines = "\n" * 6
-        file.write(empty_lines + content)
+        def get_checked_seff_result():
+            seff_result = subprocess.run(
+                seff_command,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
+            if "exit code" in str(seff_result.stdout):
+                return seff_result
+            else:
+                return None
 
-    exit_code, cpu_time, wall_time, memory_usage = parse_seff(str(result.stdout))
+        # sometimes `seff` needs a few extra seconds to realize the task has ended
+        result = wait_for(get_checked_seff_result, 30, 2)
+
+        status, exit_code, cpu_time, wall_time, memory_usage = parse_seff(
+            str(result.stdout)
+        )
+
+        if os.path.exists(exitcode_file):
+            with open(exitcode_file, "r") as f:
+                returncode = int(f.read())
+                logging.debug("Exit code in file %s: %d", exitcode_file, returncode)
+        else:
+            assert (
+                status != "COMPLETED"
+            ), "Should never happen: exit code not found, but task was reported COMPLETED."
+            logging.debug("Exit code not found in file: %s", exitcode_file)
+            returncode = 0
+
+        ret = {
+            "walltime": wall_time,
+            "cputime": cpu_time,
+            "memory": memory_usage,
+            "exitcode": ProcessExitCode.create(value=returncode),
+        }
+
+        if status != "COMPLETED":
+            ret["terminationreason"] = {
+                "OUT_OF_MEMORY": "memory",
+                "OUT_OF_ME+": "memory",
+                "TIMEOUT": "cputime",
+                "ERROR": "failed",
+                "FAILED": "killed",
+                "CANCELLED": "killed",
+            }.get(status, status)
+
+            if ret["terminationreason"] == "memory":
+                ret["memory"] = memory
+            if ret["terminationreason"] == "cputime":
+                ret["cputime"] = timelimit
+
+        # Runexec would populate the first 6 lines with metadata
+        with open(log_file, "w+") as file:
+            with open(tmp_log, "r") as log_source:
+                content = log_source.read()
+                file.write(f"{' '.join(args)}\n")
```

Comment:
Please also use the escaping here. This is intended (and really useful) for copy-pasting.

Reply:
Sure, thanks. Sometimes I feel like I keep repeating my mistakes..

Reply:
Don't feel bad! The lesson from this should only be to remember forever how important it is to have a proper way to handle command lines that does *not* need quoting/escaping, because it is clear that developers will at some point inevitably forget it. This is why I pushed for it. :-) For log output the effect will be limited, but for real execution that would be bad. Same lesson as all security-minded web developers will tell you about including user input in web pages.